README.md

```diff
@@ -12,90 +16,114 @@ Programs written for distributed execution engines usually suffer from some well
 
-Assume for a moment that `DataBag[A]` instances can be constructed in one of three ways: As the `Empty` bag, a singleton bag `Sng(x)`, or the union of two existing bags `Union(xs, ys)`. *Structural recursion* works on bags by 
+Assume for a moment that `DataBag[A]` instances can be constructed in one of three ways:
+the `Empty` bag, a singleton bag `Singleton(x)`, or the union of two existing bags `Union(xs, ys)`.
+*Structural recursion* works on bags by
 
 1. systematically deconstructing the input `DataBag[A]` instance, 
-1. replacing the constructors with corresponding user-defined functions, and 
-1. evaluating the resulting expression.
+2. replacing the constructors with corresponding user-defined functions, and 
+3. evaluating the resulting expression.
 
 Formally, the above procedure can be specified as the following second-order function called `fold`.
 
 ```scala
 def fold[B](e: B)(s: A => B, u: (B, B) => B): B = this match {
   case Empty         => e
-  case Sng(x)        => s(x)
+  case Singleton(x)  => s(x)
   case Union(xs, ys) => u(xs.fold(e)(s, u), ys.fold(e)(s, u))
 }
 ```
 
-In the above signature, `e` substitutes `Empty`, `s(x)` substitutes `Sng(x)`, and `u(xs, ys)` substitutes `Union(xs, ys)`.
+In the above signature, `e` substitutes `Empty`,
+`s(x)` substitutes `Singleton(x)`,
+and `u(xs, ys)`substitutes `Union(xs, ys)`.
 
 Various collection processing primitives can be specified as a `fold`. 
 
-```scala 
-val min = xs.fold(Int.MaxValue)(identity, Math.min(_,_))
-val max = xs.fold(Int.MinValue)(identity, Math.max(_,_))
-val sum = xs.fold(0, x => 1, _ + _)
+```scala
+val size = xs.fold(0L)(const(1L), _ + _)
+val min  = xs.fold(None)(Some(_), lift(_ min _))
+val sum  = xs.fold(0)(identity, _ + _)
 ```
 
-*Emma* offers [pre-defined aliases for common `fold` operators](emma-language/src/main/scala/eu/stratosphere/emma/macros/Folds.scala):
+*Emma* offers pre-defined aliases for common `fold` operators:
 
-Fold Alias                     | Purpose
--------------------------------|-----------------------------
-`min`, `max`, `sum`, `count` … | Aggregation
-`exists`, `forall`, …          | Existential Qualifiers
+Fold Alias                          | Purpose
+------------------------------------|-----------------------------
+`reduce`, `reduceOption`            | General
+`isEmpty`, `nonEmpty`, `size`       | Cardinality
+`min`, `max`, `top`, `bottom`       | Order-based aggregation
+`sum`, `product`                    | Numeric aggregation
+`exists`, `forall`, `count`, `find` | Predicate testing
+`sample`                            | Random sampling
 
 #### Declarative Dataflows
 
-Joins and cross products in *Emma* can be declared using `for`-comprehension syntax in a manner akin to *Select-From-Where* expressions known from SQL. 
+Joins and cross products in *Emma* can be declared using `for`-comprehension syntax
+in a manner akin to *Select-From-Where* expressions known from SQL. 
 
 ```scala
```

Comment:
More readable if the bindings of the generators and the join attributes are different

```scala
for {
  email    <- emails
  sender   <- people
  receiver <- people  
  if email.from == sender.email
  if email.to   == receiver.email
} yield (email, sender, receiver) // DataBag[(Email, Person, Person)]
```